src/Storages/ExternalStream/Kafka/Kafka.h

```diff
@@ -44,13 +43,21 @@ class Kafka final : public StorageExternalStreamImpl
     void shutdown() override {
         LOG_INFO(logger, "Shutting down Kafka External Stream");
 
-        consumer_pool->shutdown();
-        if (producer)
-            producer->shutdown();
-
         /// Must release all resources here rather than relying on the deconstructor.
         /// Because the `Kafka` instance will not be destroyed immediately when the external stream gets dropped.
-        consumer_pool.reset();
+        {
+            std::lock_guard<std::mutex> lock(consumer_mutex);
+            for (auto & consumer_ptr : consumers)
+                /// if the consumer is still running, mark it stopped
+                if (auto consumer = consumer_ptr.lock())
+                    consumer->setStopped();
+
+            std::vector<std::weak_ptr<RdKafka::Consumer>> empty_consumers;
+            consumers.swap(empty_consumers);
+        }
+
+        if (producer)
+            producer->setStopped();
```

Comment:
let's add empty line to introduce thought group

Reply:
Will fix.